firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java

```diff
@@ -185,9 +192,24 @@ private String readJsonString(int length) throws IOException {
    */
   private boolean pullMoreData() throws IOException {
     buffer.compact();
-    int read = dataReader.read(buffer);
+
+    int available = bundleInputStream.available();
+    int bytesToRead = Math.min(available, buffer.remaining());
+    // `available` is an estimation, we still try to read if the estimation is 0, to move things
+    // forward.
+    if (available == 0) {
+      bytesToRead = buffer.remaining();
+    }
```

Comment:
Can we always read buffer.remaining?

Reply:
Yes, since we are reading into the buffer directly now!